src/devices/acurite.c

```diff
@@ -713,36 +714,95 @@ static int acurite_txr_decode(r_device *decoder, bitbuffer_t *bitbuffer)
         // M = Message type
         message_type = bb[2] & 0x3f;
 
-        // tower sensor messages are 7 bytes.
-        // TODO: - see if there is a type in the message that
-        // can be used instead of length to determine type
+        // Multiple AcuRite sensors use the same basic message format,
+        // with a shared header and integrity scheme. message type is packed
+        // as the last 3 bits of the 3rd byte (bb[2]).
         if (browlen == ACURITE_TXR_BITLEN / 8) {
+            // Verify parity bits
+            // Bytes 2, 3, 4, and 5 should all have a parity bit in their MSB
+            int parity = parity_bytes(&bb[2], 4);
+            if (parity) {
+                if (decoder->verbose)
+                    bitrow_printf(bb, 7 * 8, "%s: bad parity: ", __func__);
+                continue; // return DECODE_FAIL_MIC;
+            }
```

Comment:
This whole function badly needs to be refactored because all of these if/continues are a little bit of a minefield.

Unless I'm not reading this right, because of the above, anything below this that has a message of the same length, but doesn't use the parity on bytes 2,3,4 will no long receive these rows when the values happen to fail the parity check.

Reply:
In this specific case, there are only 2 message types with this length (Tower and Leak Detector), and we have verified they both work correctly with the parity verification. If other messages/sensors were to be added in the future, they would get the parity.

If we refactor to switch on the message type before the parity, we end up repeating the parity check over the same bytes in two places. Since there are no other (currently know) sensors/messages of this length, I think it makes sense to calculate the parity first and only once (DRY).

It does make sense as a separate PR to evaluate if the parity check can be applied to ALL AcuRite sensors, like the checksum already was.

Reply:
> In this specific case, there are only 2 message types with this length (Tower and Leak Detector), and we have verified they both work correctly

Just to make sure we are on the same page, do you agree that the use of the `continue` makes it so any code below that check that wants to decode a 7 byte message might not get messages if they don't implement the same parity on bytes 2,3,4?   Can you see how that also might not be clear to the person implementing the next Acurite device?

The use of those if's and continue's were most likely my fault originally, but a bunch of devices have been added since I originally did the tower, 5n1, and 6045m. 

Also yes I agree, there are only two sensors that have declared that they use the 7 byte (56 bit) message length. But the 899 rain sensor which doesn't `#define` it's message length, the Atlas has two messages lengths, with/without lightning detector, but only one is `#define`d.  

>  I think it makes sense to calculate the parity first and only once (DRY).

Yes, that's what kept me from implementing the parity check for all, and it was a dumb move on my part. At this point I don't know whether every device covered by this decoder implements parity on all bytes except for 0,1, and N.

So I'd rather just see the parity check repeated even if it isn't DRY. 

And most importantly, if the tower and leak sensors get refactored into their own functions the way the Atlas and 6045 are we can avoid the spaghetti and unintended consequences.

Reply:
Yes, we are on the same page. The AcuRite decode really should be refactored and cleaned up a lot, but that would make this PR a lot more involved and touch devices we don't strictly need to.

I am not sure adding one more `continue` is a meaningful barrier to someone writing another future AcuRite device, since the whole routine is pretty convoluted already. (IMO it is clearer with these changes, if only just.)

I am not opposed to re-writing the AcuRite decoder, but we don't have test .cu8 files from all of the AcuRite sensors; just the 5n1, tower, and leak detector. If you have samples of the other sensors, I would be happy to rework the whole file once we know how similar/unique the protocol actually is. But I think that is best served as a separate PR, since the goal of this one is to add the leak detector + fix the tower, and based on my testing, it does both, without affecting the other sensor decodes.

Reply:
I'm not advocating refactoring the whole acurite decoder in this PR.

I've mentioned what I think needs to change in order to get this merged and it's not hard. If you want to wait, I'll add the message type check for the tower sensor and factor it out, and then the changes for this can be rebased and resubmitted.

As you've said there are other devices you haven't tested that are covered by the same decoder. Ordering of those if's and continue's is significant.  It does seem to be the case where changes to the tower sensor are necessary to avoid the false positives from the new leak sensor. But let's not make a bad situation worse.

Reply:
The changes you propose seem reasonable. I am pretty confident the changes in this PR only affect the leak detector and the tower. And, as you pointed out, any future AcuRite decoders for 7-byte messages. But we can refactor to be more explicit at the expense of some repetition.

Reply:
I've just pushed commits to break out the tower and leak detector decode into their own functions. It makes things more readable, but does have a larger overall footprint.